{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d8824d32",
   "metadata": {},
   "source": [
    "# Project 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0423d0c3",
   "metadata": {},
   "source": [
    "## Scientific Question: Why do tea leaves (Camellia sinensis) have a different caffeine content than the same amount of coffee beans (Coffea arabica)?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75668ea0",
   "metadata": {},
   "source": [
    "Caffeine is a methylxanthine alkaloid found in a variety of plants, most notably Camellia sinensis (tea) and Coffea arabica (coffee), and is favored by many due to its stimulant nature. Caffeine inhibits adenosine-mediated downregulation of CNS activity and promotes neurotransmitter release that further stimulates the CNS (https://ncithesaurus.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&ns=ncit&code=C328). Caffeine biosynthesis in both tea and coffee plants mainly follows this pathway: xanthosine → 7-methylxanthosine → 7-methylxanthine → theobromine → caffeine (https://www.researchgate.net/figure/Caffeine-biosynthetic-pathway-in-tea-plants-showing-the-genes-studied-in-this-work-IMP_fig1_5323113#:~:text=The%20main%20biosynthetic%20pathway%20is,slow%20in%20caffeine%2Daccumulating%20plants). The enzymes involved in the synthesis of caffeine are N-methyltransferases, which transfer methyl groups from S-adenosyl-L-methionine onto caffeine precursors (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1165278/pdf/biochemj00564-0092.pdf). In tea plants, tea caffeine synthase 1 (TCS1) catalyzes the final two steps of caffeine synthesis, converting 7-methylxanthine to thebromine and theobromine to caffeine. In coffee plants, the final two steps of caffeine synthesis are catalyzed by CaMXMT and CaDXMT1, respectively (https://www.pnas.org/doi/10.1073/pnas.1602575113). CaDXMT2 has been studied relatively recently and is thought to catalyze the last two steps of caffeine synthesis, just as TCS1 does.\n",
    "\n",
    "The protein sequences and structures for TCS1 and CaDXMT can be found on the UniProt database (https://www.uniprot.org/). As stated on this database's website: \n",
    "\n",
    "\"The Universal Protein Resource (UniProt) is a comprehensive resource for protein sequence and annotation data. The UniProt databases are the UniProt Knowledgebase (UniProtKB), the UniProt Reference Clusters (UniRef), and the UniProt Archive (UniParc). The UniProt consortium and host institutions EMBL-EBI, SIB and PIR are committed to the long-term preservation of the UniProt databases. \n",
    "\n",
    "UniProt is a collaboration between the European Bioinformatics Institute (EMBL-EBI), the SIB Swiss Institute of Bioinformatics and the Protein Information Resource (PIR). Across the three institutes more than 100 people are involved through different tasks such as database curation, software development and support.\n",
    "\n",
    "EMBL-EBI and SIB together used to produce Swiss-Prot and TrEMBL, while PIR produced the Protein Sequence Database (PIR-PSD). These two data sets coexisted with different protein sequence coverage and annotation priorities. TrEMBL (Translated EMBL Nucleotide Sequence Data Library) was originally created because sequence data was being generated at a pace that exceeded Swiss-Prot's ability to keep up. Meanwhile, PIR maintained the PIR-PSD and related databases, including iProClass, a database of protein sequences and curated families. In 2002 the three institutes decided to pool their resources and expertise and formed the UniProt consortium.\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7f21d74",
   "metadata": {},
   "source": [
    "## Scientific Hypothesis: If caffeine synthases, such as TCS1 and CaDXMT1/CaDXMT2, are known to be essential to the caffeine synthesis pathway and known to have similar expression levels in Camellia sinensis and Coffea arabica, respectively, then it is likely that the difference in caffeine content between the two species comes from differences in the enzyme structures."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d309fcfa",
   "metadata": {},
   "source": [
    "There were many different known and proposed caffeine synthases from a variety of species identified from the UniProt database. Protein sequences for TCS1A, TCS1E, CaDXMT1, and CaDXMT2 were identified on UniProt and saved as a fasta file that was then read in here for a pairwise alignment analysis, the first bioinformatics analysis done in this notebook.\n",
    "\n",
    "Pairwise sequence alignment is used to identify regions of similarity between two sequences (protein sequences in this case) that may indicate functional, structural and/or evolutionary relationships.\n",
    "\n",
    "Understanding the sequences of these compounds is important to contextualize the similarity of these compounds and interpreting interpreting the heat map results. The second bioinformatics analyses done in this notebook are structural bioinformatics/homology modeling.\n",
    "\n",
    "Protein structures for TCS1A, TCS1E, CaDXMT1, and CaDXMT2 were identified on SwissProt and saved as 4 .pdb files that were then read in here. 3D measurements of atom-atom distances between certain residues of TCS1/CaDXMT were inspected to show strucutral differences between the two enzymes. ***MAYBE: The residues of each enzyme were mapped to each other and the structures were also superimposed to further investigate strucutral similarities and differences***.\n",
    "\n",
    "To answer this scientific question and test my hypothesis, I searched for TCS1 and CaDXMT (3,7-dimethylxanthine N-methyltransferase) on UniProt and I then downloaded fasta file with all four sequences and a pdb file for each individual protein. That is the data I imported here."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2f6082f",
   "metadata": {},
   "source": [
    "## Part 1: Load the Packages\n",
    "\n",
    "numpy: \n",
    "\n",
    "matplotlib:\n",
    "\n",
    "nglview:\n",
    "\n",
    "biopython: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7d364641",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): ...working... done\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "\n",
      "Solving environment: ...working... done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: C:\\Users\\Luvly_000\\anaconda3\n",
      "\n",
      "  added / updated specs:\n",
      "    - biopython\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    biopython-1.79             |   py39hb82d6ee_1         2.6 MB  conda-forge\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:         2.6 MB\n",
      "\n",
      "The following packages will be UPDATED:\n",
      "\n",
      "  biopython          pkgs/main::biopython-1.78-py39h2bbff1~ --> conda-forge::biopython-1.79-py39hb82d6ee_1\n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages\n",
      "\n",
      "biopython-1.79       | 2.6 MB    |            |   0% \n",
      "biopython-1.79       | 2.6 MB    |            |   1% \n",
      "biopython-1.79       | 2.6 MB    | ###        |  31% \n",
      "biopython-1.79       | 2.6 MB    | ######5    |  65% \n",
      "biopython-1.79       | 2.6 MB    | ########## | 100% \n",
      "biopython-1.79       | 2.6 MB    | ########## | 100% \n",
      "Preparing transaction: ...working... done\n",
      "Verifying transaction: ...working... done\n",
      "Executing transaction: ...working... done\n"
     ]
    }
   ],
   "source": [
    "conda install -c conda-forge biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6a7938a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): ...working... done\n",
      "Solving environment: ...working... done\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "conda install nglview -c conda-forge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "82f8c46b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): ...working... done\n",
      "Solving environment: ...working... \n",
      "Warning: >10 possible package resolutions (only showing differing packages):\n",
      "  - r/noarch::r-lava-1.6.5-r36h6115d3f_0, r/noarch::r-numderiv-2016.8_1-r36h6115d3f_0, r/noarch::r-squarem-2017.10_1-r36h6115d3f_0, r/win-64::r-ps-1.3.0-r36h6115d3f_0\n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.  - defaults/noarch::r-squarem-2017.10_1-r36h6115d3f_0, r/noarch::r-lava-1.6.5-r36h6115d3f_0, r/noarch::r-numderiv-2016.8_1-r36h6115d3f_0, r/win-64::r-ps-1.3.0-r36h6115d3f_0\n",
      "  - defaults/noarch::r-numderiv-2016.8_1-r36h6115d3f_0, r/noarch::r-lava-1.6.5-r36h6115d3f_0, r/noarch::r-squarem-2017.10_1-r36h6115d3f_0, r/win-64::r-ps-1.3.0-r36h6115d3f_0\n",
      "  - defaults/noarch::r-numderiv-2016.8_1-r36h6115d3f_0, defaults/noarch::r-squarem-2017.10_1-r36h6115d3f_0, r/noarch::r-lava-1.6.5-r36h6115d3f_0, r/win-64::r-ps-1.3.0-r36h6115d3f_0\n",
      "  - defaults/noarch::r-numderiv-2016.8_1-r36h6115d3f_0, defaults/win-64::r-ps-1.3.0-r36h6115d3f_0, r/noarch::r-lava-1.6.5-r36h6115d3f_0, r/noarch::r-squarem-2017.10_1-r36h6115d3f_0\n",
      "  - defaults/noarch::r-numderiv-2016.8_1-r36h6115d3f_0, defaults/noarch::r-squarem-2017.10_1-r36h6115d3f_0, defaults/win-64::r-ps-1.3.0-r36h6115d3f_0, r/noarch::r-lava-1.6.5-r36h6115d3f_0\n",
      "  - defaults/win-64::r-ps-1.3.0-r36h6115d3f_0, r/noarch::r-lava-1.6.5-r36h6115d3f_0, r/noarch::r-numderiv-2016.8_1-r36h6115d3f_0, r/noarch::r-squarem-2017.10_1-r36h6115d3f_0\n",
      "  - defaults/noarch::r-squarem-2017.10_1-r36h6115d3f_0, defaults/win-64::r-ps-1.3.0-r36h6115d3f_0, r/noarch::r-lava-1.6.5-r36h6115d3f_0, r/noarch::r-numderiv-2016.8_1-r36h6115d3f_0\n",
      "  - defaults/noarch::r-lava-1.6.5-r36h6115d3f_0, defaults/win-64::r-ps-1.3.0-r36h6115d3f_0, r/noarch::r-numderiv-2016.8_1-r36h6115d3f_0, r/noarch::r-squarem-2017.10_1-r36h6115d3f_0\n",
      "  - defaults/noarch::r-lava-1.6.5-r36h6115d3f_0, r/noarch::r-numderiv-2016.8_1-r36h6115d3f_0, r/noarch::r-squarem-2017.10_1-r36h6115d3f_0, r/win-64::r-ps-1.3.0-r36h6115d3f_0\n",
      "  ... and othersdone\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "WARNING: The --force flag will be removed in a future conda release."
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  environment location: C:\\Users\\Luvly_000\\anaconda3\n",
      "\n",
      "  added / updated specs:\n",
      "    - nglview\n",
      "\n",
      "\n",
      "\n",
      "Preparing transaction: ...working... done\n",
      "Verifying transaction: ...working... done\n",
      "Executing transaction: ...working... done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "         See 'conda update --help' for details about the --force-reinstall\n",
      "         and --clobber flags.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "conda upgrade nglview --force"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b84f6d33",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Enabling notebook extension nglview-js-widgets/extension...\n",
      "      - Validating: ok\n"
     ]
    }
   ],
   "source": [
    "!jupyter-nbextension enable nglview --py --sys-prefix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fb8a4f22",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): ...working... done\n",
      "Solving environment: ...working... done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: C:\\Users\\Luvly_000\\anaconda3\n",
      "\n",
      "  added / updated specs:\n",
      "    - nglview\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    nglview-1.1.7              |             py_0         4.5 MB  bioconda\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:         4.5 MB\n",
      "\n",
      "The following packages will be SUPERSEDED by a higher-priority channel:\n",
      "\n",
      "  nglview            conda-forge::nglview-3.0.3-pyh8a188c0~ --> bioconda::nglview-1.1.7-py_0\n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages\n",
      "\n",
      "nglview-1.1.7        | 4.5 MB    |            |   0% \n",
      "nglview-1.1.7        | 4.5 MB    |            |   0% \n",
      "nglview-1.1.7        | 4.5 MB    | 7          |   7% \n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "\n",
      "nglview-1.1.7        | 4.5 MB    | #1         |  11% \n",
      "nglview-1.1.7        | 4.5 MB    | ##8        |  29% \n",
      "nglview-1.1.7        | 4.5 MB    | ####9      |  49% \n",
      "nglview-1.1.7        | 4.5 MB    | #######5   |  75% \n",
      "nglview-1.1.7        | 4.5 MB    | ########## | 100% \n",
      "nglview-1.1.7        | 4.5 MB    | ########## | 100% \n",
      "Preparing transaction: ...working... done\n",
      "Verifying transaction: ...working... done\n",
      "Executing transaction: ...working... done\n"
     ]
    }
   ],
   "source": [
    "conda install -c bioconda nglview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ff647ad8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correctly import all the packages needed \n",
    "# Includes downloading packages as seen above.\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.PDB import *\n",
    "import nglview as ngl"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c28a9ab0",
   "metadata": {},
   "source": [
    "## Part 2: Load in the data and perform Bioinformatics Analyses (Pairwise Sequence Analysis & Structural Bioinformatics)\n",
    "\n",
    "Fasta file format is a text-based format for representing either nucleotide sequences or peptide sequences, in which base pairs or amino acids are represented using single-letter codes. A sequence in FASTA format begins with a single-line description, followed by lines of sequence data. Pairwise sequence alignment returns alignments scores where a higher score indicates higher similarity between two sequences.\n",
    "\n",
    "PDB file format is a textual file format describing the three-dimensional structures of molecules held in the Protein Data Bank. These files list the atoms in each protein, and their 3D location in space. A typical PDB formatted file includes a large \"header\" section of text that summarizes the protein, citation information, and the details of the structure solution, followed by the sequence and a long list of the atoms and their coordinates. The archive also contains the experimental observations that are used to determine these atomic coordinates. Structural bioinformatics can give us quantitiative differences between 3D protein structures.\n",
    "\n",
    "In the code below, we read in protein sequences from a fasta file, and then perform pairwise sequence alignment between each pair of sequences. Use a for loop to iterate through all of the protein pairs to calculate the alignment scores to store in the matrix my_numpy_array.\n",
    "\n",
    "Later in the code, we read in protein structures from pdb files and perform structural bioinformatics as analyses (3D measurements) to compare the various caffeine synthases. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4c2c74dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the fasta file using BioPython's SeqIO built-in parse function\n",
    "# Define global variable fasta\n",
    "fasta = list(SeqIO.parse(\"DXMT1_TCS1A_DXMT2_TCS1E.fasta\", \"fasta\"))\n",
    "\n",
    "# Define the size variable to be used in the nested for loop below\n",
    "# Define global variable size\n",
    "size = len(fasta)\n",
    "\n",
    "# Initiate an empy numpy matrix that is the length of the fasta file\n",
    "# Define global variable alignments_score\n",
    "alignments_score = np.empty(shape=(size,size))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a07e144b",
   "metadata": {},
   "source": [
    "## Part 3A: Create a heatmap (plotting method) to show the alignment scores (bioinformatics method) between all compounds in the set.\n",
    "\n",
    "A heatmap is a plotting method that colors the levels of some feature that can be defined by two inputs. In this case, the two inputs are two of the potential compounds from the entire list of 4 caffeine synthases. The colors here are showing the level of protein sequence similarity between the two inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c9de7a1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sp|Q8H0D2|DXMT1_COFAR\n",
      "sp|Q8H0D2|DXMT1_COFAR\n",
      "sp|Q8H0D2|DXMT1_COFAR\n",
      "sp|Q8H0D2|DXMT1_COFAR\n",
      "sp|Q9FZN8|TCS1_CAMSI\n",
      "sp|Q9FZN8|TCS1_CAMSI\n",
      "sp|Q9FZN8|TCS1_CAMSI\n",
      "sp|Q9FZN8|TCS1_CAMSI\n",
      "sp|Q8H0D3|DXMT2_COFAR\n",
      "sp|Q8H0D3|DXMT2_COFAR\n",
      "sp|Q8H0D3|DXMT2_COFAR\n",
      "sp|Q8H0D3|DXMT2_COFAR\n",
      "sp|A0A0S2PM82|TCS3_CAMSI\n",
      "sp|A0A0S2PM82|TCS3_CAMSI\n",
      "sp|A0A0S2PM82|TCS3_CAMSI\n",
      "sp|A0A0S2PM82|TCS3_CAMSI\n"
     ]
    }
   ],
   "source": [
    "# Nested for loop to do pairwise sequence alignment for each pair of caffeine synthases\n",
    "def pairwise_seq_heatmap(my_numpy_array, size):\n",
    "    # Loop over the caffeine synthases \n",
    "    for i in range(size):\n",
    "        # Loop over the caffeine synthases again\n",
    "        for j in range(size):\n",
    "        # Define two sequences to be aligned\n",
    "        # Define local variables x and y\n",
    "            x = fasta[i].seq\n",
    "            y = fasta[j].seq\n",
    "            # Code check below\n",
    "            print(fasta[i].id)\n",
    "\n",
    "        # Use of built-in pairwise2 method gloaballxx (from biopython)\n",
    "        # Reads in two sequences (nucleotide or amino acid) and returns the\n",
    "        # sequence alignments between the two sequences with scores\n",
    "        # Get a list of the global alignments between the two sequences x and y\n",
    "        # No parameters. Identical characters have score of 1, else 0.\n",
    "        # No gap penalties.\n",
    "        # Update input array my_numpy_array\n",
    "            my_numpy_array[i,j] = int(pairwise2.align.globalxx(x, y, score_only=True))\n",
    "\n",
    "    return my_numpy_array\n",
    "\n",
    "# Call the pairwise_seq_heatmap function defined above with arguments alignments_score and size\n",
    "# Define global variable My_alignment_Scores\n",
    "My_alignment_Scores = pairwise_seq_heatmap(alignments_score, size)\n",
    "# Check that the for loop correctly created the matrix\n",
    "#print(My_alignment_Scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e3466a12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAATwAAAD8CAYAAADqmhgGAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAatUlEQVR4nO3df7BdZX3v8feHGAF/1MgNavglsQSvwGiQTMQ6KI14CQy3sR1rwx9KrTNpHLyjM3VGqDNaneHe3rbauV6saToyDXccMXMRzNBQGhlRmWkIP4Q0IaEEcGpKLhmCBFJoNOd87h/rOXZ72HufFdY6Z++d/Xkxa87aaz17Pc8hk2+etZ71fB/ZJiJiHBw36AZERMyVBLyIGBsJeBExNhLwImJsJOBFxNhIwIuIsfGKJl+WdBLwbeBM4CfAh23/rEu5nwDPAxPAEdvLmtQbEfFyNO3hXQPcaXsJcGf53Mtv2l6aYBcRg9I04K0CNpT9DcAHG14vImLWqMlMC0nP2l7Q8flntl/fpdwTwM8AA39te32fa64B1gC8+lW64D+f9cqX3b5h9eiO1wy6CbPmjeccGnQTZsX/2/GqQTdhVvw7/8bPfVhNrnHpb77aB56ZqFX2/u2H77C9skl9Tcz4DE/S94A3dTn1uaOo5z22n5T0BmCLpN22f9itYAmG6wGWveMEb7vj9KOoZjRcdtZvDLoJs+Yzm+4ZdBNmxZ+d9fZBN2FW3DP5vcbXePqZCe6547RaZecvemxh4wobmDHg2b6k1zlJT0laZHufpEXA/h7XeLL83C/pFmA50DXgRcSoMROeHHQjamn6DG8TcFXZvwr47vQCkl4t6bVT+8B/AXY0rDcihoSBSVxrG7SmAe9PgQ9IehT4QPmMpFMkbS5l3gjcLekhYBvwd7b/vmG9ETFEJmv+N2iN3sOzfQB4f5fjTwKXl/3HgXc0qScihpcxvxiRW9pGAS8iwsDEENyu1pGAFxGNtfF8TtIJVIOZx1PFpv9r+wuSvg28tRRbADxre6mkM4FdwCPl3Fbba/vVkYAXEY0YmGgnc/phYIXtQ5LmUz37v932700VkPRl4GDHdx6zvbRuBQl4EdFYG0/wXM2CmHpzfX7ZfhlJJQn4MLDi5daRbCkR0YgxEzU3YKGk+zq2NZ3XkjRP0oNU7/Rusd35JvtFwFO2H+04tljSjyX9QNJFM7U1PbyIaMSGX9S/o326XwIR2xPAUkkLgFsknWd76r3dK4FvdRTfB5xh+4CkC4BbJZ1r+7le108PLyIaEhM1t7psPwvcBawEkPQK4Heo0tFNlTlcXo3D9v3AY8DZ/a6bgBcRjRiYdL2tH0knl54dkk4ELgF2l9OXALtt751Wfl7ZfwuwBHi8Xx25pY2Ixo6m99bHImBDCWLHARtt31bOreZXb2cB3gt8SdIRquTCa20/06+CBLyIaKR68bh5wLO9HTi/x7nf73LsZuDmo6kjAS8iGjHwC4/G07EEvIhoxIiJERkOSMCLiMYm3cozvFmXgBcRjbT1DG8uJOBFRENiIs/wImIcVBmPE/AiYgzY4ueeN+hm1JKAFxGNTY7IM7xW+qGSVkp6RNIeSdd0OS9JXy3nt0t6Zxv1RsTgVYMWx9XaBq1xD69MA/ka1SI+e4F7JW2y/XBHscuo5rktAd4FfL38jIiRNzqDFm20cjmwx/bjtn8O3ASsmlZmFXCjK1uBBWUd24gYcVODFnW2QWujBacCP+34vLccO9oyETGiJqxa26C1MWjR7beYngimTpmqYJUBdQ3AGadmTCVi2BnxC4/G39U2enh7gdM7Pp8GPPkyygBge73tZbaXnfyfRmOoO2KcjdKgRRstuBdYImmxpFdS5a3aNK3MJuCjZbT2QuCg7X0t1B0RA2bq3c4eE7e0to9I+iRwBzAPuMH2Tklry/l1wGbgcmAP8ALwsab1RsTwGIYBiTpaufG2vZkqqHUeW9exb+DqNuqKiOFiMzKvpYzGk8aIGFrVoMVoPG8fjbAcEUOtjUELSSdI2ibpIUk7JX2xHP8TSf8q6cGyXd7xnWvLDK5HJF06UzvTw4uIRozaSgB6GFhh+5Ck+cDdkm4v5/7S9l90FpZ0DtUg6bnAKcD3JJ1d1rbtKj28iGisjR5emYl1qHycX7Z+izuuAm4q69M+QTUourxfHQl4EdFItS7tcbU2YKGk+zq2NZ3XkjRP0oPAfmCL7XvKqU+WxCM3SHp9OXbUM7gS8CKiITFRcwOenppYULb1nVeyPWF7KdXkhOWSzqNKNvLrwFJgH/DlX1b8Un2X+07Ai4hGqmUa59Xaal/Tfha4C1hp+6kSCCeBv+E/bltrz+CakoAXEY3YOppb2p4knSxpQdk/EbgE2D0ts9JvAzvK/iZgtaTjJS2mSj+3rV8dGaWNiMZaevF4EbCh5Ng8Dtho+zZJ/0fSUqrO5E+APwQoM7o2Ag8DR4Cr+43QQgJeRDRU5cNr/lqK7e3A+V2Of6TPd64DrqtbRwJeRDQ0OhmPE/AiopHqtZTBZ0KpIwEvIhoZpbm0CXgR0dhYpYeKiPFVpYfKLW1EjIk8w4uIsVBlS8ktbUSMgWpqWQJeRIyF0enhtdJKSStLxtE9kq7pcv5iSQc7MpZ+vo16I2I4TKJa26A17uGVeW9fAz5Alb3gXkmbbD88reiPbF/RtL6IGC7jNkq7HNhj+3EASTdRZSKdHvCO2qM7XsNlZ/1G08sMnf+x8/uDbsKsufZt7xt0E2bFiu0HBt2EWbHrw33n2tc2Tre0dbOOvrssznG7pHN7XUzSmqlsqD/3v7fQvIiYTVNrWtTZBq2NHl6drKMPAG8ui3NcDtxKlbvqpV+sMqCuB3jdvIV9s5dGxOAZODJGPbwZs47afm5qcY6yaPd8SQtbqDsihkAbCUDnQhstuBdYImmxpFdSLZu2qbOApDdJUtlfXuo9Nh+KRIybmrezx8Qtre0jkj4J3AHMA24omUjXlvPrgA8Bn5B0BHgRWG07t6sRx4C2EoDOhVZePC63qZunHVvXsX89cH0bdUXE8BmG3lsdg7+pjoiRNpUAtOktraQTJG0rb3PslPTFcvzPJe0u69Le0rHQz5mSXuyY0LCubwVkallENGTEkclW+k6HgRXlbY75wN2Sbge2ANeWx2f/E7gW+Gz5zmNlHdta0sOLiMbamFrmyqHycX7ZbPsfbB8px7dSvQnysiTgRUQzbueWFqqpqpIeBPYDW2zfM63IHwC3d3xeLOnHkn4g6aKZrp9b2oho5CgX8Vko6b6Oz+vLZIPqWtW6skvLc7pbJJ1neweApM9RrT/7zVJ8H3CG7QOSLgBulXSu7ed6VZ6AFxGNHUXAe9r2spkK2X5W0l3ASmCHpKuAK4D3T73SZvsw1XM/bN8v6THgbOC+7lfNLW1ENGTExORxtbZ+JJ3cMQJ7InAJsFvSSqpBit+y/cK08vPK/luopqs+3q+O9PAiorGWXjxeBGwoQew4YKPt2yTtAY4HtpQJW1ttrwXeC3ypTGiYANbafqZfBQl4EdGI3c6Lx7a3A+d3OX5Wj/I3AzcfTR0JeBHRmEdkpkUCXkQ0NByJAepIwIuIxtLDi4ixYMPEZAJeRIyJsUoPFRHjy+SWNiLGRgYtImKMjEr+8gS8iGhsVG5pW5lLK+kGSfsl7ehxXpK+KmlPyVr6zjbqjYjBq0Zpm8+lnQttteBvqbIa9HIZ1cTeJcAa4Ost1RsRQ8Cutw1aKwHP9g+BfpN2VwE3loymW4EFkha1UXdEDJ6tWtugzdUzvFOBn3Z83luO7ZteUNIaql4gJ+jVc9K4iHj5zHAEszrm6qa62/+Nrh1c2+ttL7O97JU6YZabFRFtcM1t0Oaqh7cXOL3j82nAk3NUd0TMJoNHZGrZXPXwNgEfLaO1FwIHbb/kdjYiRtNYPcOT9C3gYqoFOvYCX6BaYg3b64DNwOXAHuAF4GNt1BsRw2EYRmDraCXg2b5yhvMGrm6jrogYLplLGxHjw8CIBLzBv/ocESOvjRePJZ0gaZukhyTtlPTFcvwkSVskPVp+vr7jO9eWGVyPSLp0pnYm4EVEQ8KT9bYZHAZW2H4HsBRYWQY5rwHutL0EuLN8RtI5wGrgXKqZXn81tWxjLwl4EdFcCy/ilZlYh8rH+WUz1UytDeX4BuCDZX8VcJPtw7afoBoUXd6vjgS8iGjGR/VaykJJ93VsazovJWmepAeB/cAW2/cAb5x6ja38fEMp3msGV08ZtIiI5uq/lvK07WU9L2NPAEslLQBukXRen2vVnsE1JT28iGiBam712H4WuIvq2dxTU8lGys/9pdhRz+BKwIuI5iZrbn1IOrn07JB0InAJsJtqptZVpdhVwHfL/iZgtaTjJS2mSj+3rV8duaWNiGbaew9vEbChjLQeB2y0fZukfwQ2Svo48C/A7wLY3ilpI/AwcAS4utwS95SAFxGNtTG1zPZ24Pwuxw8A7+/xneuA6+rWkYAXEc2N01zaiBhzIzK1LAEvIhpTengRMRYsGJEEoAl4EdFcengRMTYS8CJibCTgRcRYGLcEoJJukLRf0o4e5y+WdFDSg2X7fBv1RsRwkOttg9ZWD+9vgeuBG/uU+ZHtK1qqLyKGyRAEszraWsTnh5LObONaETF6hqH3VsdcPsN7t6SHqNK3fMb2zm6FSkLANQBvOOUVfObue+awiXPj2re9b9BNmDW3P7510E2YFZeesnTQTZgVz7lvRvT6xukZXg0PAG8uuer/N3Brr4K219teZnvZ605q6Q8jImZP3fTuQ9ALnJOAZ/u5qVz1tjcD8yUtnIu6I2IOjEjAm5NbWklvAp6ybUnLqQLtgbmoOyJmn2ZI7jksWgl4kr4FXEy1QMde4AtUKw5hex3wIeATko4ALwKr7TYyaEXEUBiRv81tjdJeOcP566leW4mIY8ywvGNXR2ZaRERzGaWNiLHRwqCFpNMlfV/SLkk7JX2qHP92xyytn5R1a5F0pqQXO86tm6mZ6eFFRGMt3dIeAf7I9gOSXgvcL2mL7d/7ZT3Sl4GDHd95zPbSuhUk4EVEM25nlNb2PmBf2X9e0i7gVKpVyZAk4MPAipdbR25pI6K5+re0CyXd17Gt6Xa5MlX1fKBzqtVFVK+3PdpxbLGkH0v6gaSLZmpmengR0Vz9W9qnbS/rV0DSa4CbgU/bfq7j1JXAtzo+7wPOsH1A0gXArZLOnfadX5GAFxGNtfVaiqT5VMHum7a/03H8FcDvABdMHbN9GDhc9u+X9BhwNnBfr+vnljYihkJ5RvcNYJftr0w7fQmw2/bejvInS5pX9t8CLAEe71dHengR0Vw7Pbz3AB8B/mnq1RPgj8v8+9X86u0swHuBL5UZXBPAWtvP9KsgAS8immlvlPZuoOsbzLZ/v8uxm6luf2tLwIuI5jK1LCLGgchc2ogYJwl4ETEWki0lIsbKOCUAjYjxlh5eRIyPBLyIGAtDskBPHY2nlvVK2jetjCR9VdIeSdslvbNpvRExPKbSvM+0DVobPbxeSfse7ihzGdU8tyXAu4Cvl58RcSwYgmBWR+Menu19th8o+88DU0n7Oq0CbnRlK7BA0qKmdUfEcNBkvW3QWs2W0iNpH1QB8Kcdn/fy0qA4dY01U8kBDz4z0WbzImI21E3+OQS9wNYCXp+kfdB9QnDXX9/2etvLbC973Unz2mpeRMwSHcU2aG0txN01aV+HvcDpHZ9PA55so+6IGAJD0Huro41R2n5J+6ZsAj5aRmsvBA6WBTsi4hgwTqO0XZP2AWcA2F4HbAYuB/YALwAfa6HeiBgWQxDM6mgc8Pol7esoY+DqpnVFxBBqKQHoXMhMi4hobkR6eFnEJyIaa+MZXq9ZW5L+RNK/SnqwbJd3fOfaMoPrEUmXztTO9PAiorl2enhdZ22Vc39p+y86C0s6h2pxn3OBU4DvSTrbds8XeNPDi4jG2ujh1Zy11WkVcJPtw7afoBoUXd6vjgS8iGjGVAlA62ywcGomVdnWdLtkl1lbnyyJR26Q9PpyrPYMrikJeBHRyNQiPjV7eE9PzaQq2/qXXO+ls7a+Dvw6sBTYB3y5o+rp+vYjE/AiormW5tJ2m7Vl+ynbE7Yngb/hP25bj3oGVwJeRDQmu9bW9xo9Zm1Ny6z028COsr8JWC3peEmLqdLPbetXR0ZpI6KZ9jKh9Jq1daWkpaWWnwB/CGB7p6SNwMNUI7xX9xuhhQS8iGhBG/Nk+8za2tznO9cB19WtIwEvIhrL1LKIGB8jMrUsAS8imhmS1E91JOBFRHMJeBExDqZePB4FCXgR0ZgmRyPiJeBFRDNDsiJZHQl4EdHYqLyW0sYiPl2T9k0rc7Gkgx0J/D7ftN6IGCIjsi5tGz28rkn7bD88rdyPbF/RQn0RMWTGZtCiLLe4r+w/L2kqad/0gBcRxyIDMyQGGBatPsPrkrSv07slPUSVvuUztnf2uMYaYA3ACbyKPzvr7W02cSis2H5g0E2YNZeesnTQTZgV//2Jvkk4RtZH/+u/tXKdUXmG11rA65K0r9MDwJttHyoLcNxKlcrlJUpCwPUAv6aTRuOfjYgxNkrv4bWSD69b0r5Otp+zfajsbwbmS1rYRt0RMWB2/W3A2hil7Zq0b1qZN5VySFpe6j127+sixkwbi/jMhTZuaXsl7TsDwPY64EPAJyQdAV4EVttDEO4joh0j8re5jVHaXkn7OstcD1zftK6IGE7D0HurIzMtIqIZAxOjEfGyiE9ENNbGM7xes7Yk/bmk3WVd2lskLSjHz5T0YscMrnUztTMBLyKaa2eUdmrW1tuAC4GrJZ0DbAHOs/124J+Bazu+85jtpWVbO1MFCXgR0VgbPTzb+2w/UPafB3YBp9r+B9tHSrGtVOvPviwJeBHRTN3EAVXAWyjpvo5tTbdL9pm19QfA7R2fF0v6saQfSLpopqZm0CIiGhGg+oMWT9te1vd6PWZtSfoc1W3vN8uhfcAZtg9IugC4VdK5XWZ6/VICXkQ0ppZeq+01a0vSVcAVwPun3uG1fRg4XPbvl/QYcDZwX6/rJ+BFRDMt5brrNWtL0krgs8D7bL/Qcfxk4BnbE5LeQjU///F+dSTgRURDrc2T7TVr66vA8cCWMkN1axmRfS/wpTKDawJYa/uZfhUk4EVEY23MtOgza2tzj/I3U93+1paAFxHNjcjU+AS8iGjGRzVKO1AJeBHR3GjEuwS8iGiurddSZlsCXkQ0l4AXEWPBwLgt4hMR40k4t7QRMUYmR6OL18YiPidI2ibpoZK074tdykjSVyXtKUn83tm03ogYElO3tHW2AWujh3cYWFHWnJ0P3C3pdttbO8pcRjXPbQnwLuDr5WdEHANG5Za2cQ/PlUPl4/yyTf/tVwE3lrJbgQWSFjWtOyKGxLisSwsgaV6Z7Lsf2GJ7etK+U4GfdnzeW45FxMgbo4W4AWxP2F5KlXp5uaTzphXpNiG4628vac1UNtRfVKmuImKYTa1aVmcbsFZTvNt+FrgLWDnt1F7g9I7PpwFP9rjGetvLbC+bz/FtNi8iZonsWtugtTFKe3LHsmknApcAu6cV2wR8tIzWXggctL2vad0RMSRG5Ja2jVHaRcAGSfOoAuhG27dJWgtgex1VPqvLgT3AC8DHWqg3IoaBgcnBB7M6Ggc829upVheafnxdx76Bq5vWFRHDaDh6b3VkpkVENDciAS/r0kZEMwYmJuttfUg6XdL3Je0qs7Y+VY6fJGmLpEfLz9d3fOfaMoPrEUmXztTUBLyIaMjgyXpbf0eAP7L9NuBC4GpJ5wDXAHfaXgLcWT5Tzq0GzqV6M+SvylhCTwl4EdFcC6O0tvfZfqDsPw/sopqgsArYUIptAD5Y9lcBN9k+bPsJqkHR5f3qyDO8iGjm6EZpF0rqXCh7ve310wtJOpNqMPQe4I1Tr7HZ3ifpDaXYqUDnnP0ZZ3Al4EVEc/UHLZ62vaxfAUmvoVp+8dO2nytr0XYt2q0l/a6dW9qIaK6lF49LxqWbgW/a/k45/NRUspHyc385XnsG15QEvIhoxoaJiXpbH6q6ct8Adtn+SsepTcBVZf8q4Lsdx1dLOl7SYqr0c9v61ZFb2ohorp338N4DfAT4p5J9CeCPgT8FNkr6OPAvwO9WVXqnpI3Aw1QjvFfb7htVE/AiorkWAp7tu+n+XA7g/T2+cx1wXd06EvAioiGPz1zaiBhzBs/8UvFQSMCLiOZmmDY2LBLwIqIZe2SWaUzAi4jmRiRbSgJeRDTm9PAiYjwkAWhEjItxSvEeEePNgGeYNjYs2li17ARJ2yQ9VLKUfrFLmYslHZT0YNk+37TeiBgSbi0B6Kxro4d3GFhh+1DJdHC3pNttb51W7ke2r2ihvogYMh6XW9qyItmh8nF+2Ubjt4+IdgxB760OuYXRlZJH/n7gLOBrtj877fzFVDmu9lLlq/qM7Z09rrUGWFM+vhV4pHED61kIPD1Hdc2l/F6jZy5/tzfbPrnJBST9PVWb63ja9som9TXRSsD75cWkBcAtwH+zvaPj+K8Bk+W293Lgf5UFOYaGpPtmysQ6ivJ7jZ5j+XcbtFYTgNp+FriLagWhzuPP2T5U9jcD8yXV/RchIqIVbYzSnlx6dkg6EbgE2D2tzJtKNlMkLS/1Hmhad0TE0WhjlHYRsKE8xzsO2Gj7NklrAWyvAz4EfELSEeBFYLXbvJdux0tWTjpG5PcaPcfy7zZQrT7Di4gYZlnEJyLGRgJeRIyNsQ94klZKekTSHknXDLo9bZF0g6T9knbMXHp0SDpd0vcl7SpTGT816Da1oc4UzWhurJ/hlYGWfwY+QPVS9L3AlbYfHmjDWiDpvVQzYG60fd6g29OWshDzItsPSHot1QvvHxz1P7PyFsOrO6doAp/qMkUzGhj3Ht5yYI/tx23/HLgJWDXgNrXC9g+BZwbdjrbZ3mf7gbL/PLALOHWwrWrOlUzRnGXjHvBOBX7a8Xkvx8BfnnEh6UzgfOCeATelFZLmlQWo9wNbbB8Tv9cwGfeA123R3/yrOgIkvYZqfvanbT836Pa0wfaE7aXAacByScfMo4hhMe4Bby9wesfn06iSG8QQK8+4bga+afs7g25P23pN0Yzmxj3g3QsskbRY0iuB1cCmAbcp+igP978B7LL9lUG3py11pmhGc2Md8GwfAT4J3EH18Htjr7RVo0bSt4B/BN4qaa+kjw+6TS15D/ARYEVHBu3LB92oFiwCvi9pO9U/xFts3zbgNh1zxvq1lIgYL2Pdw4uI8ZKAFxFjIwEvIsZGAl5EjI0EvIgYGwl4ETE2EvAiYmz8f1L8mCdBROJTAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Create subplots\n",
    "# Define global variables fig and ax\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# Actually show the heatplot with matplotlib built-in imshow function\n",
    "plt.imshow(alignments_score)\n",
    "\n",
    "# Construct the heat map to show the axes\n",
    "# Define global variable im\n",
    "im = ax.imshow(alignments_score)\n",
    "\n",
    "# Add a colorbar to the right of the heatmap\n",
    "# Define global variable cbar\n",
    "cbar = ax.figure.colorbar(im)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eec54d4",
   "metadata": {},
   "source": [
    "## Part 4.1: Analysis of the results (heat map)\n",
    "\n",
    "The plot above shows the similarities between protein sequences of, in order, TCS1A, DXMT1, TCS1E, and DXMT2. All four enzymes were downloaded from the UniProt. Caffeine synthases that are more similar to one another have a higher pairwise alignment score (closer to 380 & yellow) and those that are less similar to one another have a lower pairwise alignment score (closer to 200 & purple). Each cell in the matrix represents a pairwise alignment score score. The central diagonal all yellow because these are comparing the similarity between a compound and itself. The two TCS1 enzymes (TCS1A and TCS1E) have a high similarity with a pairwise alignment score of around 360. The two CaDXMT enzymes (DXMT1 and DXMT2) have a slightly lower sequence similarity with a pairwise alignment score of around 340. However, pairs of enzymes between species (i.e. TCS1A and DXMT1 or 2) have very low pairwise alignment scores, around 200, indicating around 50% difference between these sequences (200 similarities out of about 400 amino acids total)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a88e54af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Distance: 22.520088\n"
     ]
    }
   ],
   "source": [
    "# Create a PDBParser object in order to read in pdb files\n",
    "# Define global variable parser\n",
    "parser = PDBParser()\n",
    "\n",
    "# Use the parser to read in the TCS1A pdb file\n",
    "# Define global variable tcs1a_pdb\n",
    "tcs1a_pdb = parser.get_structure('TCS1A', 'AF-Q9FZN8-F1-model_v2.pdb')\n",
    "# Use nglview's built-in function show_biopython to get the protein structure\n",
    "# Define global variable tcs1a_structure\n",
    "tcs1a_structure = ngl.show_biopython(tcs1a_pdb)\n",
    "\n",
    "# Retrieve the model information from the pdb file\n",
    "# Define global variable model\n",
    "model = tcs1a_pdb[0]\n",
    "# Retrieve the chain information from the model\n",
    "# Define global variable chain\n",
    "chain = model[\"A\"]\n",
    "# Retrieve the first instance of GLY residue\n",
    "# Define global variable GLY1\n",
    "GLY1 = chain[7] #hard coded ***fix\n",
    "# Retrieve the second instance of GLY residue\n",
    "# Define global variable GLY2\n",
    "GLY2 = chain[18] #hard coded ***fix\n",
    "# Retrieve the location of the alpha carbon of GLY1\n",
    "# Define global variable ca1\n",
    "ca1 = GLY1[\"CA\"]\n",
    "# Retrieve the location of the alpha carbon of GLY2\n",
    "# Define global variable ca2\n",
    "ca2 = GLY2[\"CA\"]\n",
    "\n",
    "#res_list = Selection.unfold_entities(tcs1a_pdb, \"R\") #maybe use to find first and second instances of GLY;\n",
    "# maybe should use instances of proline instead bc it's known to change structures (bc bulky)\n",
    "\n",
    "# Find the distance between two atoms\n",
    "# Define global variable distance\n",
    "distance = ca1 - ca2\n",
    "# Check that a distance has been returned\n",
    "#print ('Distance: ' + str(distance))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "87caeb2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3371265fda4745469d816e0e514c51d7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# View image of 3D structure of TCS1A\n",
    "tcs1a_structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "5740a8e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Distance: 3.8684144\n"
     ]
    }
   ],
   "source": [
    "# Use the parser to read in the DXMT2 pdb file\n",
    "# Define global variable DXMT2\n",
    "dxmt2_pdb = parser.get_structure('DXMT2', 'AF-Q8H0D3-F1-model_v2.pdb')\n",
    "# Use nglview's built-in function show_biopython to get the protein structure\n",
    "# Define global variable dxmt2_structure\n",
    "dxmt2_structure = ngl.show_biopython(dxmt2_pdb)\n",
    "\n",
    "# Retrieve the model information from the pdb file\n",
    "# Define global variable model2\n",
    "model2 = dxmt2_pdb[0]\n",
    "# Retrieve the chain information from the model\n",
    "# Define global variable chain2\n",
    "chain2 = model[\"A\"]\n",
    "# Retrieve the first instance of GLY residue in DXMT2\n",
    "# Define global variable GLY3\n",
    "GLY3 = chain[11] #hard coded ***fix\n",
    "# Retrieve the second instance of GLY residue in DXMT2\n",
    "# Define global variable GLY4\n",
    "GLY4 = chain[12] #hard coded ***fix\n",
    "# Retrieve the location of the alpha carbon of GLY3\n",
    "# Define global variable ca3\n",
    "ca3 = GLY3[\"CA\"]\n",
    "# Retrieve the location of the alpha carbon of GLY4\n",
    "# Define global variable ca4\n",
    "ca4 = GLY4[\"CA\"]\n",
    "\n",
    "#res_list = Selection.unfold_entities(dxmt2_pdb, \"R\")\n",
    "\n",
    "# Find the distance between two atoms; effectively the distance between the first to GLY\n",
    "# Define global variable distance2\n",
    "distance2 = ca3 - ca4\n",
    "print ('Distance: ' + str(distance2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "d80a3e01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "30a2c1bc7ebe4927abd9424bc8c7f965",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# View image of 3D structure of TCS1A\n",
    "dxmt2_structure"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "150ea9f7",
   "metadata": {},
   "source": [
    "## Part 4.2: Analysis of the results (structural bioinformatics)\n",
    "\n",
    "If we look at the structure of the two caffeine synthases above (TCS1A and DXMT2), the lack of similarity seen in the heat map makes sense in that caffeine synthases from Camellia sinensis are very different from caffeine synthases from Coffea arabica. Not only are the protein sequences different, but so are the protein structures. The folding and conformations of TCS1A and DXMT2 are quite different. Thus, assuming that expression levels of these two enzymes are similar in their respective plants and that they both play the same role in the caffeine biosynthesis pathway (catalyzing the final two steps), then my hypothesis is likely to be correct and the difference in caffeine content between Camellia sinensis and Coffea arabica is due to strucutral differences of their caffeine synthases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3cee9c7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
